dynfu::tracking_lost_error

jordan-heemskerk opened the pull request:
Closes #105

jordan-heemskerk committed:
dynfu::tracking_lost_error

jordan-heemskerk committed:
dynfu::tracking_lost_error: add test

 - Document previous test cases to make a little more clear

RobertLeahy commented on include/dynfu/pose_estimation_pipeline_block.hpp:
```diff
@@ -82,6 +82,20 @@ namespace dynfu {
 		
```
This file doesn't `#include <stdexcept>` which is [where `std::runtime_error` comes from](http://en.cppreference.com/w/cpp/error/runtime_error).

RobertLeahy commented on include/dynfu/pose_estimation_pipeline_block.hpp:
```diff
@@ -82,6 +82,20 @@ namespace dynfu {
+
+		public:
+
+			tracking_lost_error();
```
I would normally not define a constructor for my error type and instead:

`using std::runtime_error::runtime_error;`

which simply pulls in [all of `std::runtime_error`'s constructors](http://en.cppreference.com/w/cpp/error/runtime_error).

This has the advantage of allowing the code throwing the exception to specify a more specific error message.  Perhaps some `dynfu::pose_estimation_pipeline_block` implementations decide they lost tracking at 10%, some at 25%, some at a certain number.

Note that if you want to allow the throwing code to specify a custom error message and provide a canned error message for convenience you can:

```
tracking_lost_error();
using std::runtime_error::runtime_error;
```

Then you can accept the canned error message (throw a default constructed `dynfu::tracking_lost_error`) or provide your own (by passing a string to the constructor).

jordan-heemskerk replied:
Let's just use std::runtime_error constructors

RobertLeahy commented on src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp:
```diff
@@ -95,7 +95,7 @@ namespace dynfu {
 		if (correspondences.size() < 0.1*frame_width_*frame_height_) {
 			// I think this means we have blown our tracking... what to do...
-			throw std::runtime_error("Less than 10% of correspondences were accepted, tracking lost");
+			throw tracking_lost_error();
```
Whenever I default construct something in this way I use so-called [list initialization](http://en.cppreference.com/w/cpp/language/list_initialization) so the line would look like this:

```
throwing tracking_lost_error{};
```

This has the benefit of being more general as you'll never run into the so-called [most vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse).

Note that it doesn't change the meaning whatsoever in this situation it's just a style nitpick.

jordan-heemskerk replied:
Ah, but now I have a string in the constructor, so I should still use () right?

RobertLeahy replied:
That's what I would do.  Everyone's going to disagree on how strongly they prefer list initialization.

RobertLeahy commented on src/pose_estimation_pipeline_block.cpp:
```diff
@@ -6,5 +6,6 @@ namespace dynfu {
 	
 	pose_estimation_pipeline_block::~pose_estimation_pipeline_block () noexcept {	}
 	
+	tracking_lost_error::tracking_lost_error () : std::runtime_error("The pose estimation pipeline block has lost tracking.") {}
```
I don't think error messages should have trailing periods.

jordan-heemskerk replied:
OK

RobertLeahy commented on src/test/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp:
```diff
@@ -152,7 +152,37 @@ SCENARIO_METHOD(fixture, "A kinect_fusion_eigen_pose_estimation_pipeline_block u
 			
+			THEN("A tracking failure occurs") {
+				
+					CHECK_THROWS_AS(kfepeb(t, t2, k, t_gk_minus_one), dynfu::tracking_lost_error);
```
I think this line is indented one level too many.

jordan-heemskerk replied:
Indeed

jordan-heemskerk committed:
dynfu::tracking_lost_error:

  dynfu::tracking_lost_error -> dynfu::pose_estimation_pipeline_block::tracking_lost_error
